cladecombiner/aggregator.py

```diff
@@ -0,0 +1,278 @@
+from abc import ABC, abstractmethod
+from collections.abc import Iterable
+from copy import copy
+from dataclasses import dataclass
+from warnings import warn
+
+from .ordered_taxon import OrderedTaxon
+from .taxon import Taxon
+from .taxonomy_scheme import PhylogeneticTaxonomyScheme
+from .utils import table
+
+
+@dataclass
+class TaxonMapping:
+    """
+    Class for tracking proposed step in aggregation process.
+    """
+
+    map: dict[Taxon, Taxon]
+    "from : to dictionary of the mapping"
+    done: dict[Taxon, bool]
+    "to : add_to_stack dictionary tracking if the aggregated taxon should be added back to the stack"
+
+
+class Aggregator(ABC):
+    """
+    A partially-abstract base class for aggregation.
+
+    Descendants need only implement next_agg_step() to create a usable aggregation object.
+    """
+
+    def __init__(
+        self,
+        in_taxa: Iterable[Taxon],
+    ):
+        """
+        Constructor for aggregation infrastructure shared by all subclasses.
+
+        Parameters
+        ----------
+        in_taxa : Iterable[Taxon]
+            The taxa we wish to aggregate.
+        """
+        self.input_taxa = list(in_taxa)
+        "The taxa we wish to aggregate."
```

Comment:
sidebar: I'm used to having comments above a thing:
```
# the next line does xyz
x = y * z
```
but this is in the opposite order, with a string:
```
x = y * z
"xyz operation"
```

Is this a Python thing? docstring? dataclass?

Reply:
Bingo, it's a docstrings thing.

This is a comment in the traditional sense
```
# the next line does xyz
x = y * z
```


This is a docstring
```
self.x = y * z
"my x value"
```
Which is why you'll note that these only show up under attribute assignments.